File: android/app/src/main/java/com/llmhub/llmhub/mimobot/pipeline/VoicePipeline.kt
package com.llmhub.llmhub.mimobot.pipeline

import android.util.Log
import com.llmhub.llmhub.data.LLMModel
import com.llmhub.llmhub.embedding.RagService
import com.llmhub.llmhub.inference.InferenceService
import com.llmhub.llmhub.mimobot.audio.AudioSink
import com.llmhub.llmhub.mimobot.speech.SpeechToText
import com.llmhub.llmhub.mimobot.speech.Tts
import com.llmhub.llmhub.websearch.DuckDuckGoSearchService
import com.llmhub.llmhub.websearch.SearchIntentDetector
import com.llmhub.llmhub.websearch.WebSearchService
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Voice loop wiring: STT → LLM → TTS → AudioSink.
 *
 * Transport-agnostic: pass a [SpeakerSink][com.llmhub.llmhub.mimobot.audio.SpeakerSink]
 * for local dev mode, or a future `BleAudioSink` (Opus-encode + BLE write) for
 * the real device.
 *
 * Reuses existing LLM-Hub services unchanged:
 *   - [InferenceService]  → any of the 4 inference backends
 *   - [RagService]        → optional, passes document context into the prompt
 *   - [WebSearchService]  → optional, gated by [SearchIntentDetector]
 */
class VoicePipeline(
    private val scope: CoroutineScope,
    private val inference: InferenceService,
    private val model: LLMModel,
    private val stt: SpeechToText,
    private val tts: Tts,
    private val sink: AudioSink,
    private val chatId: String = "mimobot",
    private val rag: RagService? = null,
    private val webSearch: WebSearchService? = DuckDuckGoSearchService(),
    private val webSearchEnabled: Boolean = true,
) {

    enum class State { IDLE, LISTENING, THINKING, SPEAKING }

    private val _state = MutableStateFlow(State.IDLE)
    val state: StateFlow<State> = _state.asStateFlow()

    private val _lastTranscript = MutableStateFlow("")
    val lastTranscript: StateFlow<String> = _lastTranscript.asStateFlow()

    private val _lastResponse = MutableStateFlow("")
    val lastResponse: StateFlow<String> = _lastResponse.asStateFlow()

    private var currentTurn: Job? = null
    @Volatile private var cancelRequested = false

    /** Run one listen → think → speak turn. Safe to call only when state == IDLE. */
    fun startTurn() {
        if (currentTurn?.isActive == true) return
        cancelRequested = false
        currentTurn = scope.launch(Dispatchers.Default) {
            try {
                runTurn()
            } catch (t: Throwable) {
                Log.e(TAG, "turn failed", t)
            } finally {
                _state.value = State.IDLE
            }
        }
    }

    /** Cancel an in-flight turn. Mic, LLM stream, and TTS are all torn down. */
    fun cancel() {
        cancelRequested = true
        stt.cancel()
        tts.stop()
        sink.stop()
        currentTurn?.cancel()
        currentTurn = null
        _state.value = State.IDLE
    }

    private suspend fun runTurn() {
        _state.value = State.LISTENING
        val transcript = stt.recognizeTurn().trim()
        if (cancelRequested) return
        _lastTranscript.value = transcript
        if (transcript.isBlank()) {
            Log.d(TAG, "empty transcript, abandoning turn")
            return
        }

        _state.value = State.THINKING
        val prompt = augmentPrompt(transcript)

        _state.value = State.SPEAKING
        _lastResponse.value = ""
        sink.start()

        // Sentence channel decouples LLM streaming from TTS synthesis so the first
        // audio frame starts rendering before the model is done generating.
        val sentences = Channel<String>(Channel.UNLIMITED)
        val speakJob = scope.launch(Dispatchers.IO) {
            for (sentence in sentences) {
                if (cancelRequested) break
                try {
                    tts.speakToPcm(sentence).collect { frame ->
                        if (cancelRequested) return@collect
                        sink.write(frame)
                    }
                } catch (t: Throwable) {
                    Log.w(TAG, "tts emit failed: ${t.message}")
                }
            }
        }

        try {
            val buf = StringBuilder()
            inference.generateResponseStream(prompt, model).collect { token ->
                if (cancelRequested) return@collect
                buf.append(token)
                _lastResponse.value = buf.toString()
                while (true) {
                    val end = findSentenceEnd(buf)
                    if (end < 0) break
                    val sentence = buf.substring(0, end + 1).trim()
                    buf.delete(0, end + 1)
                    if (sentence.isNotEmpty()) sentences.send(sentence)
                }
            }
            val tail = buf.toString().trim()
            if (tail.isNotEmpty() && !cancelRequested) sentences.send(tail)
        } finally {
            sentences.close()
            speakJob.join()
            sink.stop()
        }
    }

    /** Find the index of the first sentence-ending char in [buf], or -1. */
    private fun findSentenceEnd(buf: StringBuilder): Int {
        // Allow a short minimum so "Hi." still gets spoken, but prefer longer chunks
        // so we don't chop in the middle of a short phrase.
        if (buf.length < 12) return -1
        for (i in buf.indices) {
            val c = buf[i]
            if (c == '.' || c == '!' || c == '?' || c == '\n') return i
        }
        return -1
    }

    private suspend fun augmentPrompt(userQuery: String): String {
        val sb = StringBuilder()

        rag?.let { r ->
            val chunks = r.searchRelevantContext(
                chatId = chatId,
                query = userQuery,
                maxResults = 3,
                relaxedLexicalFallback = false,
                queryEmbedding = null,
            )
            if (chunks.isNotEmpty()) {
                sb.append("CONTEXT FROM YOUR DOCUMENTS:\n")
                chunks.forEach { sb.append("- ").append(it.content).append('\n') }
                sb.append('\n')
            }
        }

        if (webSearchEnabled && webSearch != null &&
            SearchIntentDetector.needsWebSearch(userQuery)
        ) {
            try {
                val q = SearchIntentDetector.extractSearchQuery(userQuery)
                val results = webSearch.search(q, maxResults = 3)
                if (results.isNotEmpty()) {
                    sb.append("CURRENT WEB SEARCH RESULTS:\n")
                    results.forEach { r ->
                        sb.append("- ").append(r.title).append(": ").append(r.snippet).append('\n')
                    }
                    sb.append('\n')
                }
            } catch (t: Throwable) {
                Log.w(TAG, "web search failed: ${t.message}")
            }
        }

        sb.append("system: You are a small voice companion. Answer in one or two short sentences. Never output Markdown, code blocks, or emoji.\n\n")
        sb.append("user: ").append(userQuery)
        return sb.toString()
    }

    companion object { private const val TAG = "MimoVoicePipeline" }
}

File: ios/LLMHub/Sources/LLMHub/MimoBot/VoicePipeline.swift
import Foundation
import SwiftUI

/// Voice loop wiring: STT → LLM → TTS → AudioSink.
///
/// Transport-agnostic: pass a `SpeakerSink` for local dev mode or a future
/// `BLEAudioSink` for the real device. Reuses `LLMBackend.shared` for
/// generation (which itself wraps RunAnywhere + llama.cpp / Apple Foundation).
@MainActor
final class VoicePipeline: ObservableObject {

    enum State: String { case idle, listening, thinking, speaking }

    /// A single completed exchange.
    struct Turn: Identifiable, Equatable {
        let id = UUID()
        let user: String
        let assistant: String
    }

    @Published private(set) var state: State = .idle
    @Published private(set) var lastTranscript: String = ""
    @Published private(set) var lastResponse: String = ""
    @Published private(set) var history: [Turn] = []

    private var stt: SpeechToText
    private var tts: TTS
    private var sink: AudioSink
    private let rag: RagService?
    private let webSearchEnabled: Bool
    private let chatId: String
    private let maxHistoryTurns: Int

    private var currentTurn: Task<Void, Never>?

    init(
        stt: SpeechToText,
        tts: TTS,
        sink: AudioSink,
        chatId: String = "mimobot",
        rag: RagService? = nil,
        webSearchEnabled: Bool = true,
        maxHistoryTurns: Int = 6
    ) {
        self.stt = stt
        self.tts = tts
        self.sink = sink
        self.chatId = chatId
        self.rag = rag
        self.webSearchEnabled = webSearchEnabled
        self.maxHistoryTurns = maxHistoryTurns
    }

    /// Hot-swap the TTS impl. Stops any in-flight synth on the old one first.
    func setTTS(_ newTTS: TTS) {
        tts.stop()
        tts = newTTS
    }

    /// Wipe the conversation memory. Called from the test screen's Clear button.
    func clearHistory() { history = [] }

    func startTurn() {
        if currentTurn != nil { return }
        lastResponse = ""
        currentTurn = Task { [weak self] in
            guard let self else { return }
            await self.runTurn()
            await MainActor.run { self.currentTurn = nil; self.state = .idle }
        }
    }

    func cancel() {
        stt.cancel()
        tts.stop()
        sink.stop()
        currentTurn?.cancel()
        currentTurn = nil
        state = .idle
    }

    private func runTurn() async {
        state = .listening
        let transcript = await stt.recognizeTurn(languageHint: "en-US")
        if Task.isCancelled { return }
        lastTranscript = transcript
        if transcript.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return }

        state = .thinking
        let prompt = await augmentPrompt(userQuery: transcript)

        state = .speaking
        sink.start()

        // Decouple LLM streaming from TTS synthesis so the first audio frame
        // starts rendering before the model is finished generating.
        let (sentences, sentenceCont) = AsyncStream<String>.makeStream()

        let speakTask = Task { [weak self] in
            guard let self else { return }
            for await sentence in sentences {
                if Task.isCancelled { break }
                for await frame in self.tts.speakToPCM(text: sentence, language: "en-US") {
                    if Task.isCancelled { break }
                    await self.sink.write(frame)
                }
            }
            await MainActor.run { self.sink.stop() }
        }

        // Bridge the callback-driven LLMBackend.generate API into an AsyncStream
        // so the rest of the loop can iterate over cumulative-text snapshots
        // without any @escaping mutable captures.
        let (tokenStream, tokenCont) = AsyncStream<String>.makeStream()
        let generateTask = Task.detached {
            do {
                try await LLMBackend.shared.generate(
                    prompt: prompt,
                    systemPrompt: nil,
                    onUpdate: { cumulativeText, _, _ in
                        tokenCont.yield(cumulativeText)
                    }
                )
            } catch {
                print("⚠️ VoicePipeline: LLM generate failed: \(error)")
            }
            tokenCont.finish()
        }

        var buf = ""
        var sent = 0  // chars of cumulative text already chunked
        var fullReply = ""  // running total for committing to history
        for await cumulative in tokenStream {
            if Task.isCancelled { break }
            if cumulative.count > sent {
                let idx = cumulative.index(cumulative.startIndex, offsetBy: sent)
                buf += String(cumulative[idx...])
                sent = cumulative.count
                lastResponse = cumulative
                fullReply = cumulative
                while let end = Self.findSentenceEnd(buf) {
                    let sentence = String(buf[..<buf.index(after: end)])
                        .trimmingCharacters(in: .whitespacesAndNewlines)
                    buf.removeSubrange(..<buf.index(after: end))
                    if !sentence.isEmpty { sentenceCont.yield(sentence) }
                }
            }
        }
        await generateTask.value

        let tail = buf.trimmingCharacters(in: .whitespacesAndNewlines)
        if !tail.isEmpty { sentenceCont.yield(tail) }
        sentenceCont.finish()
        await speakTask.value

        // Commit this turn to history once we've finished speaking it.
        let trimmedReply = fullReply.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmedReply.isEmpty, !Task.isCancelled {
            var updated = history + [Turn(user: transcript, assistant: trimmedReply)]
            if updated.count > maxHistoryTurns {
                updated.removeFirst(updated.count - maxHistoryTurns)
            }
            history = updated
        }
    }

    private static func findSentenceEnd(_ s: String) -> String.Index? {
        guard s.count >= 12 else { return nil }
        return s.firstIndex { c in c == "." || c == "!" || c == "?" || c == "\n" }
    }

    private func augmentPrompt(userQuery: String) async -> String {
        var sb = ""

        if let rag {
            let chunks = await rag.search(
                chatId: chatId,
                query: userQuery,
                queryEmbedding: nil,
                maxResults: 3
            )
            if !chunks.isEmpty {
                sb += "CONTEXT FROM YOUR DOCUMENTS:\n"
                for c in chunks { sb += "- \(c.content)\n" }
                sb += "\n"
            }
        }

        // TODO(augment-web): iOS web-search lives inside ChatScreen.swift as
        // a fileprivate actor; hoist it before wiring here. See TODO in
        // ios/LLMHub/Sources/LLMHub/MimoBot/README-style TODOs.
        _ = webSearchEnabled

        sb += "system: You are a small voice companion. Answer in one or two short sentences. Never output Markdown, code blocks, or emoji.\n\n"

        // Replay conversation memory so the model sees previous turns. Format
        // matches what LLMBackend.generate's prompt parser expects ("user: …"
        // / "assistant: …" lines) — Harmony rebuilding stays correct because
        // both pieces follow the same convention.
        for turn in history {
            sb += "user: \(turn.user)\n"
            sb += "assistant: \(turn.assistant)\n"
        }

        sb += "user: \(userQuery)"
        return sb
    }
}
